snapcraft/pluginhandler.py

```diff
@@ -301,7 +359,21 @@ def _new_clean(self, project_staged_state=None,
         if not project_stripped_state:
             project_stripped_state = {}
 
-        self._clean_steps(project_staged_state, project_stripped_state, step)
+        try:
+            self._clean_steps(project_staged_state, project_stripped_state,
+                              step)
+        except MissingState:
+            # If one of the step cleaning rules is missing state, it must be
+            # running on the output of an old Snapcraft. In that case, if we
+            # were specifically asked to clean that step we need to fail.
+            # Otherwise, just clean like the old Snapcraft did, and blow away
+            # the entire part directory.
+            if step:
+                raise
+
```

Comment:
no need for a warning or debug in here? if !step, it could mean that we have the output of an old snapcraft, or that there is a bug we missed. logging something could be useful.